File: src/nac/token/token_cursor.cpp
#include "sodium/nac/token/token_cursor.h"

#include <optional>

#include "sodium/nac/token/token.h"
#include "sodium/nac/token/token_buffer.h"

namespace sodium {

TokenCursor::TokenCursor(const TokenBuffer &token_buffer)
        : token_buffer_(std::make_unique<TokenBuffer>(token_buffer)),
          index_(-1) {}

std::optional<Token> TokenCursor::next() {
    return token_buffer_->get(++index_);
}

std::optional<Token> TokenCursor::peek_ahead(size_t n) const {
    return token_buffer_->get(index_ + n);
}

size_t TokenCursor::index() const {
    return index_;
}

} // namespace sodium

File: nac/src/token/token_cursor.rs
use super::{Token, TokenStream};

pub struct TokenCursor<'ts> {
    tokens: &'ts TokenStream,
    index: usize,
}

impl<'ts> TokenCursor<'ts> {
    pub fn new(tokens: &'ts TokenStream) -> Self {
        Self { tokens, index: 0 }
    }

    pub fn next(&mut self) -> Option<&'ts Token> {
        self.tokens.vec().get(self.index).map(|token| {
            self.index += 1;
            token
        })
    }
}
